File: typescript/solved/unique-binary-search-trees-ii.ts
/**
 * https://leetcode.com/problems/unique-binary-search-trees-ii/
 */
export { generateTrees, TreeNode };

function generateTrees(n: number): Array<TreeNode | null> {
  function dfs(start: number, end: number) {
    if (end < start) return [null];
    if (end === start) return [new TreeNode(end)];

    const trees: TreeNode[] = [];

    for (let i = start; i <= end; i++) {
      const leftTrees = dfs(start, i - 1);
      const rightTrees = dfs(i + 1, end);

      for (const left of leftTrees) {
        for (const right of rightTrees) {
          trees.push(new TreeNode(i, left, right));
        }
      }
    }

    return trees;
  }

  return dfs(1, n);
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: python/today7/unique_binary_search_trees_ii.py
# https://leetcode.com/problems/unique-binary-search-trees-ii/description/
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def generate(left: int, right: int) -> List[Optional[TreeNode]]:
            if left > right:
                return [None]

            res = []
            for val in range(left, right+1):
                for left_tree in generate(left, val-1):
                    for right_tree in generate(val+1, right):
                        root = TreeNode(val, left_tree, right_tree)
                        res.append(root)
            return res
        return generate(1, n)
            
